### PR TITLE
Automatic passivation for sharding

### PR DESCRIPTION
**Pull request by ismaelhamed**
Port [#25765](https://github.com/akka/akka/pull/25765)

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding.Tests/InactiveEntityPassivationSpec.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+            var probe = new TestProbe(Sys, new XunitAssertions());
+            var settings = ClusterShardingSettings.Create(Sys);
+            var region = ClusterSharding.Get(Sys).Start(
+                "myType",
+                Entity.Props(probe.Ref),
+                settings,
+                _extractEntityId,
+                _extractShard,
+                new LeastShardAllocationStrategy(10, 3),
```
Comment (ismaelhamed) [inline src/contrib/cluster/Akka.Cluster.Sharding.Tests/InactiveEntityPassivationSpec.cs]: `ClusterSharding.DefaultShardAllocationStrategy()` seems to be accessible from here in the JVM (see [#L69](https://github.com/johanandren/akka/blob/27296b4c808f967f6d636ef44c0a958fc93e0a44/akka-cluster-sharding/src/test/scala/akka/cluster/sharding/InactiveEntityPassivationSpec.scala#L69))

File: src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs [hunk 2]
```diff
@@ -617,14 +624,28 @@ internal static class Shards
             }
         }
 
-        public static void PassivateCompleted<TShard>(this TShard shard, Shard.EntityStopped evt) where TShard: IShard
+        private static void PassivateIdleEntities<TShard>(this TShard shard) where TShard : IShard
```
Comment (ismaelhamed) [inline src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs]: My first implementation was easier to the eye and more in line with the JVM, but probably less _performant_:

```csharp
var deadline = DateTime.Now.Ticks - shard.Settings.PassivateIdleEntityAfter.Ticks;
var refsToPassivate = shard.LastMessageTimestamp.Where(s => s.Value < deadline).Select(s => shard.RefById[s.Key]).ToArray();
if (refsToPassivate.Any())
{
    shard.Log.Debug($"Passivating [{refsToPassivate.Length}] idle entities");
    refsToPassivate.ForEach(entity => Passivate(shard, entity, shard.HandOffStopMessage));
}
```

Comment (Horusiath) [reply src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs]: 👍

File: src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs [hunk 3]
```diff
@@ -660,26 +681,61 @@ internal static class Shards
             }
             else
             {
-                if (shard.MessageBuffers.TryGetValue(id, out var buffer))
+                // TODO
```
Comment (Horusiath) [inline src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs]: TODO what?

Comment (ismaelhamed) [reply src/contrib/cluster/Akka.Cluster.Sharding/Shard.cs]: Nada. Missed this one before pushing...

**Comment by Aaronontheweb** [comment]
LGTM

**Pull request merged**